### PR TITLE
Add `center_and_scale()` function for preprocessing vector data

### PR DESCRIPTION
**Pull request by jjallaire**
cc @topepo @terrytangyuan @kevinykuo

### CODE REVIEW

File: R/preprocessing.R [hunk 1]
```diff
@@ -738,6 +738,59 @@ flow_images_from_directory <- function(
...
+#' train_data <- center_and_scale(train_data, values)
+#' test_data <- center_and_scale(train_data, values)
+#' }
+#'
+#' @name center_and_scale
+#' @export
+fit_center_and_scale <- function(x) {
+  
+  # find mean and std
+  mean <- apply(x, 2, mean)
```
Comment (topepo) [inline R/preprocessing.R]: I would add `na.rm  = TRUE` for this and the std dev calculation

**Comment by topepo** [comment]
You might note somewhere that some people normalize predictors for neural networks by standardizing them to be on some range such as `[-1, 1]` and this this is not that. 

Looks good to me.

File: R/preprocessing.R (hunk 1 above)
Comment (jjallaire) [reply R/preprocessing.R]: Although the interface to `scale` which we call later requires a vector of the same number of rows as `x` -- if we did the na.rm when taking the mean and sd wouldn't that potentially make the vector the wrong size? 

In that case it seems like either:

1) The user should do an `na.rm` on the entire matrix before preprocessing; or

2) We should do an na.rm on the entire matrix in both the `fit_center_and_scale` and the `center_and_scale` functions.

Does this make sense or am I missing something?

**Comment by jjallaire** [comment]
I'm going to merge this b/c it's referenced in an article I have on the web. Will make any required changes based on discussion (e.g. the `na.rm` one discussed above) directly on master.

**Comment by kevinykuo** [comment]
Late to the party, but I'm wondering if the naming/interface could be potentially confusing. Right now there are a couple meanings for `fit_`, one is for fitting models with data, e.g. `fit_generator()`, and the other is updating a transformer (`fit_text_tokenizer()`). This PR introduces a third pattern, which is in spirit similar to `fit_text_tokenizer()` but the interface is a bit different.

Would it make sense to do something like
- Having a `matrix_scaler()` function that instantiates a transformer, here we can pass in things like `na_rm` or `center` and `scale`.
- Having a `fit_matrix_scaler()` (or a `fit()` generic) that has signature `function(object, data)`, somewhat like `fit_text_tokenizer()`.
- Having a `transform_matrix()` (or a `transform()` generic) that has signature `function(object, data)`, analogous to `texts_to_matrix()`

So the user would write something like

```
scaler <- matrix_scaler() %>%
  fit_matrix_scalar(train_data)
train_data <- transform_matrix(scaler, train_data)
test_data <- transform_matrix(scalaer, test_data)
```

Just throwing some ideas out there, we can totally forget about this if it'd introduce too much complexity.

**Comment by topepo** [comment]
I think that the idea was to keep it small in scope and simple (otherwise they could just use `caret::preProcess` or `recipes`). 

If the wording is still in question, I'm cool with `transform_matrix` but `matrix_scaler` and `fit_matrix_scalar` don't sit right with me. Maybe use the verb "estimate" if an alternative is needed.

**Comment by kevinykuo** [comment]
Lol I managed to spell scaler three different ways in that snippet. Yeah the main issue I had was the inconsistencies with `fit_` but it's fine if it doesn't bother anyone else.

**Comment by jjallaire** [comment]
The idea was to be consistent with `fit_text_tokenizer` and `fit_image_data_generator` where you "fit" some context that is later applied (those in term come from `fit` methods in Keras e.g. `model.fit`, `text_tokenizer.fit`, and `data_generator.fit`. In the later two cases it implies "fit the behavior / coefficients of this object based on some data", that's kind of what I was going for here.

As for "transform" that as Max says starts to sound like a general purpose transformation mechanism, which we have well covered in caret and recipes (and with that sort of name it might underwhelm expectations as Scikit learn has a whole "transformers" construct that does much more (as with caret and recipes).

I don't however want to put something in here that's just plain wrong or off or strikes an odd chord with anyone. As I write this and think about how narrow this particular method is, I'm wondering if we aren't better served by the original code:

```r
mean <- apply(train_data, 2, mean)
train_data <- scale(train_data, center = mean, scale = FALSE)
std <- apply(train_data, 2, sd)
train_data <- scale(train_data, center = FALSE, scale = std)
test_data <- scale(test_data, center = mean, scale = std)
```

For sake of comparison here's the Python code it is equivalent to:

```python
mean = train_data.mean(axis=0)
train_data -= mean
std = train_data.std(axis=0)
train_data /= std
test_data -= mean
test_data /= std
```

Yes it's a bit verbose compared to Scikit learn or Caret but it does actually show the user more clearly what's going on, and in the process teaches them about the R `apply` function which is (along with `sweep`) kind of indispensable for non-trivial preprocessing (at least in base R).

**Comment by kevinykuo** [comment]
I, too, am starting to feel like perhaps we don't really need the helper functions. Also, we can get by with one fewer line of code:

```
mean <- apply(train_data, 2, mean)
std <- apply(train_data, 2, sd)
train_data <- scale(train_data, center = mean, scale = std)
test_data <- scale(test_data, center = mean, scale = std)
```
`apply()` should be fair game to R users and `scale()` is simple enough.

**Comment by jjallaire** [comment]
Indeed one less line! Much nicer.
